src/goto-cc/linker_script_merge.h

```diff
@@ -0,0 +1,201 @@
+/// \file linker_script_merge.h
+/// \brief Merge linker script-defined symbols into a goto-program
+/// \author Kareem Khazem <karkhaz@karkhaz.com>
+
+#ifndef CPROVER_GOTO_CC_LINKER_SCRIPT_MERGE_H
+#define CPROVER_GOTO_CC_LINKER_SCRIPT_MERGE_H
+
+#include <functional>
+
+#include <util/cout_message.h>
+#include <util/json.h>
+
+#include "compile.h"
+#include "gcc_cmdline.h"
+
+/// \brief Patterns of expressions that should be replaced
+///
+/// Each instance of this class describes an expression 'shape' that should be
+/// replaced with a different expression. The intention is that if some
+/// expression matches the pattern described by this replacement_predicatet
+/// (i.e. replacement_predicatet::match() returns true), then that expression
+/// should be replaced by a new expression.
+class replacement_predicatet
+{
+public:
+  replacement_predicatet(
+      const std::string &description,
+      const std::function<const symbol_exprt(const exprt&)> inner_symbol,
+      const std::function<bool(const exprt&)> match)
+    : description(description),
+      inner_symbol(inner_symbol),
+      match(match)
+  {}
+
+  /// \brief a textual description of the expression that we're trying to match
+  const std::string description;
```

Comment:
Prefer not to use const data members: www.drdobbs.com/the-problem-with-const-data-members/184403306

This goes for the whole PR.